### PR TITLE
[VARC] Variable Composites table

### PR DESCRIPTION
**Pull request by behdad**
Based on https://github.com/harfbuzz/boring-expansion-spec/issues/103#issuecomment-1856325577

### CODE REVIEW

**Comment by behdad** [comment]
This is mostly steady now.  I'll go ahead and write a spec for it.

What is left to do is:
- Rip out the glyf1 VarComposite glyphs,
- Hook VARC up in ttGlyphSet so we can draw.

**Comment by behdad** [comment]
I should also somehow figure out how to do the `VARC` glyph loading lazy.

**Comment by behdad** [comment]
TODO:
- [ ] Make VARC table lazy-load glyphs.
- [ ] Subsetter
- [ ] Instancer

**Comment by behdad** [comment]
I can use some directions here. The `multiVarStore` and `varStore` share some similar code. Should I try to subclass them from a common class?

It will be more pronounced when adding subsetting to `multiVarStore`.

**Comment by behdad** [comment]
> I can use some directions here. The `multiVarStore` and `varStore` share some similar code. Should I try to subclass them from a common class?
> 
> It will be more pronounced when adding subsetting to `multiVarStore`.

I managed to share some code by monkey-patching for now.

**Comment by behdad** [comment]
This is ready for review. Thanks in advance @anthrotype @justvanrossum

**Comment by behdad** [comment]
> This is ready for review. Thanks in advance @anthrotype @justvanrossum

It's terse but here's something of a spec for this:
https://github.com/harfbuzz/boring-expansion-spec/blob/main/varc.md

**Comment by behdad** [comment]
I probably should move some of the new code from `otTables` to a side module.

File: Lib/fontTools/misc/transform.py [hunk 1]
```diff
@@ -423,6 +423,39 @@ class DecomposedTransform:
     tCenterX: float = 0
     tCenterY: float = 0
 
+    def __eq__(self, other):
```
Comment (anthrotype) [inline Lib/fontTools/misc/transform.py:426]: this is a dataclass so it already implements `__eq__` and `__ne__` by default, no? eq=True by default
why do you need to override them?

`__bool__` probably is not implemented for you so you may want to add that manually, but the other two should already be there.

File: Lib/fontTools/subset/__init__.py [hunk 2]
```diff
@@ -3297,33 +3363,33 @@ def _closure_glyphs(self, font):
                     self.glyphs.add(font.getGlyphName(i))
                 log.info("Added first four glyphs to subset")
 
-        if self.options.layout_closure and "GSUB" in font:
-            with timer("close glyph list over 'GSUB'"):
+        if "MATH" in font:
```
Comment (anthrotype) [inline Lib/fontTools/subset/__init__.py:3366]: the swapping of GSUB and MATH closure order should arguably go to a separate PR.. unless is really required for the changes at hand

File: Lib/fontTools/ttLib/tables/_g_v_a_r.py [hunk 3]
```diff
@@ -116,32 +124,26 @@ def decompile(self, data, ttFont):
...
+            {glyphs[gid]: _decompileVarGlyph for gid in range(self.glyphCount)}
+        )
+        l.reverseGlyphMap = ttFont.getReverseGlyphMap()
+        l._glyf = glyf
+        l._gvar = self
+        l.gvarData = data
+        l.offsetToData = offsetToData
+        l.sharedCoords = sharedCoords
+        l.axisTags = axisTags
+        l.tableFormat = self.flags & 1
```
Comment (anthrotype) [inline Lib/fontTools/ttLib/tables/_g_v_a_r.py:144]: why turn `_decompileVarGlyph` into a kind-of instance method (that takes self but is defined as standalone function) and not keep it as a closure as it was before? you wouldn't need to set all these attribute on the LazyDict instance, they'd be captured from the outer namespace.

File: Lib/fontTools/ttLib/tables/otBase.py [hunk 4]
```diff
@@ -1150,7 +1150,7 @@ def fromXML(self, name, attrs, content, font):
         except KeyError:
             raise  # XXX on KeyError, raise nice error
         value = conv.xmlRead(attrs, content, font)
-        if conv.repeat:
+        if conv.repeat is not None:
```
Comment (anthrotype) [inline Lib/fontTools/ttLib/tables/otBase.py:1153]: What prompted this? can repeat ever be 0?

Comment (behdad) [reply Lib/fontTools/ttLib/tables/otBase.py:1153]: I'm (ab)using a conv.repeat of `""` to mean it's a list that is handled by manual code for decompile/compile purposes.

Comment (anthrotype) [reply Lib/fontTools/ttLib/tables/otBase.py:1153]: perhaps that warrants a comment..

**Comment by anthrotype** [comment]
> The multiVarStore and varStore share some similar code. Should I try to subclass them from a common class?
> It will be more pronounced when adding subsetting to multiVarStore.
> I managed to share some code by monkey-patching for now.

It may be cleaner to subclass from a common base class, or try to factor out the shared code into separate objects or helpers and compose them accordingly to better clarify what is shared/different between the two.
It's not immediatly obvious what the effect of the line `import fontTools.varLib.varStore  # For monkey-patching` at the top of multiVarStore.py is.

**Comment by anthrotype** [comment]
I'm looking at the ttx dump of VARC table as implemented in this PR and I wonder if we could improve it a bit, make more human friendly.
For example, the way TupleValues are written out as concatenation of the tuple deltas for each region: in the old (item) VarStore, each row of deltas would contain the same number of deltas, one per region, and would look more like an actual table; whereas each multi-item's length now is variable and equals number of regions x number of values per tuple; sometimes it can get quite long and wrap to the new line.. I wonder if we could split this long concatenated list of numbers into the tuples that comprise it, one per line, so each line doesn't get exceedingly long and the grouping is explicit. I haven't fully thought this through..

**Comment by anthrotype** [comment]
the `VarComponent` element of the ttx dump can also get quite long and wide with all the optional attributes , I wonder if it would look better to have separate sub-elements for each.

So in stead of this:

```xml
<VarComponent index="0" glyphName="glyph00006" flags="3967" axisIndicesIndex="8" axisValues="(3621, 836, 9093, 13959, 13795, 5800, 1540, 0, 16384)" axisValuesVarIndex="262144" transformVarIndex="262145" translateX="-456.0" translateY="558.0" rotation="-231.0205" scaleX="0.64" scaleY="0.95" tCenterX="973.0" tCenterY="10.0"/>
```

Something like this:

```xml
<VarComponent index="0">
  <glyphName value="glyph00006"/>
  <flags="3967"/><!-- do we actually need this? -->
  <axisIndicesIndex value="8"/>
  <axisValues value="(3621, 836, 9093, 13959, 13795, 5800, 1540, 0, 16384)"/>
  <axisValuesVarIndex value="262144"/>
  <transformVarIndex value="262145"/>
  <translateX value="-456.0"/>
  <translateY value="558.0"/>
  <rotation value="-231.0205"/>
  <scaleX value="0.64"/>
  <scaleY value="0.95"/>
  <tCenterX value="973.0"/>
  <tCenterY value="10.0"/>
</VarComponent>
```

**Comment by behdad** [comment]
> ```
>   <flags="3967"/>
> ```

Thanks Cosimo. Yes, the flags are needed mostly to know which transform components get encoded, because they have have variations, so we cannot simply encode any non-default transform component.

**Comment by behdad** [comment]
> the `VarComponent` element of the ttx dump can also get quite long and wide with all the optional attributes , I wonder if it would look better to have separate sub-elements for each.
> 
> So in stead of this:
> 
> ```
> <VarComponent index="0" glyphName="glyph00006" flags="3967" axisIndicesIndex="8" axisValues="(3621, 836, 9093, 13959, 13795, 5800, 1540, 0, 16384)" axisValuesVarIndex="262144" transformVarIndex="262145" translateX="-456.0" translateY="558.0" rotation="-231.0205" scaleX="0.64" scaleY="0.95" tCenterX="973.0" tCenterY="10.0"/>
> ```
> 
> Something like this:
> 
> ```
> <VarComponent index="0">
>   <glyphName value="glyph00006"/>
>   <flags="3967"/>
>   <axisIndicesIndex value="8"/>
>   <axisValues value="(3621, 836, 9093, 13959, 13795, 5800, 1540, 0, 16384)"/>
>   <axisValuesVarIndex value="262144"/>
>   <transformVarIndex value="262145"/>
>   <translateX value="-456.0"/>
>   <translateY value="558.0"/>
>   <rotation value="-231.0205"/>
>   <scaleX value="0.64"/>
>   <scaleY value="0.95"/>
>   <tCenterX value="973.0"/>
>   <tCenterY value="10.0"/>
> </VarComponent>
> ```

Done.

**Comment by behdad** [comment]
> For example, the way TupleValues are written out as concatenation of the tuple deltas for each region: in the old (item) VarStore, each row of deltas would contain the same number of deltas, one per region, and would look more like an actual table; whereas each multi-item's length now is variable and equals number of regions x number of values per tuple; sometimes it can get quite long and wrap to the new line.. I wonder if we could split this long concatenated list of numbers into the tuples that comprise it, one per line, so each line doesn't get exceedingly long and the grouping is explicit. I haven't fully thought this through..

The VarStores are inherently not human-friendly, so I'm not sure how important this is. Currently there is no custom `toXML` involved, it's all automatic from:
```

    (       
        "MultiVarData", 
        [ 
            ("uint8", "Format", None, None, "Set to 1."), 
            ("uint16", "VarRegionCount", None, None, ""), 
            ("uint16", "VarRegionIndex", "VarRegionCount", 0, ""), 
            ("TupleList", "Item", "", 0, ""), 
        ], 
    ), 
```
If I scratch my head enough I can implement what you suggest...

**Comment by Lorp** [comment]
I'm a bit surprised to see `axisValues` as int16 rather than as float representing F2DOT14. Do we do this anywhere else in TTX?

**Comment by behdad** [comment]
> I'm a bit surprised to see `axisValues` as int16 rather than as float representing F2DOT14. Do we do this anywhere else in TTX?

That's a good point. I'll see what I can do.

**Comment by behdad** [comment]
> I'm a bit surprised to see `axisValues` as int16 rather than as float representing F2DOT14. Do we do this anywhere else in TTX?

So if I keep these in F2DOT14 in the Python objects, that's fine, just more work when applying variations to them.

**Comment by behdad** [comment]
> I'm a bit surprised to see `axisValues` as int16 rather than as float representing F2DOT14. Do we do this anywhere else in TTX?

Fixed.

File: Tests/ttLib/data/varc-ac00-ac01-500upem.ttx [hunk 5]
```diff
@@ -332,7 +332,7 @@
           <glyphName value="glyph00008"/>
           <flags value="0x12"/>
           <axisIndicesIndex value="2"/>
-          <axisValues value="(3537,)"/>
+          <axisValues value="['0.2159']"/>
```
[the remark this answers is not in the document]
Comment (behdad) [reply Tests/ttLib/data/varc-ac00-ac01-500upem.ttx:335]: Ah yeah. Sorry about that.

[the remark this answers is not in the document]
Comment (behdad) [reply Tests/ttLib/data/varc-ac00-ac01-500upem.ttx]: Fixed.

[the remark this answers is not in the document]
Comment (behdad) [reply Tests/ttLib/data/varc-ac00-ac01-500upem.ttx]: I suppose I should change it back to be a tuple.

[the remark this answers is not in the document]
Comment (anthrotype) [reply Tests/ttLib/data/varc-ac00-ac01-500upem.ttx]: > I suppose I should change it back to be a tuple.

what's the difference if you do so? immutability or what?

[the remark this answers is not in the document]
Comment (behdad) [reply Tests/ttLib/data/varc-ac00-ac01-500upem.ttx]: Let me check what the varstore outputs.... Looks like we output lists in other places. I'll revert it back.

[the remark this answers is not in the document]
Comment (anthrotype) [reply Tests/ttLib/data/varc-ac00-ac01-500upem.ttx]: that's why I was asking.. I saw list elsewhere and was wondering why this axisValues should be tuples

[the remark this answers is not in the document]
Comment (behdad) [reply Tests/ttLib/data/varc-ac00-ac01-500upem.ttx]: Looks like we write other things as lists. Reverted this to be list as well.

**Comment by anthrotype** [comment]
yeah that was my thinking too. For all the optional values that are behind a flag, their mere presence entails the respective flag `HAVE_{SOMETHING}` is set. They should not be omitted from the ttx dump if their value is equal to the default, only if their optional flag is unset.
I think there is still one flag 0 `RESET_UNSPECIFIED_AXES` that is standalone and not related to any of the optional fields; the `GID_IS_24BIT` can also be inferred by the size of the gid value.
Since there's at least one flag that cannot be inferred we should keep the flags in the ttx dump, but perhaps we could clear the bits that can be computed (like we do for glyf components flags as well)

File: Lib/fontTools/ttLib/ttGlyphSet.py [hunk 6]
```diff
@@ -285,6 +278,78 @@ def draw(self, pen):
...
+                        pen.addVarComponent(
+                            comp.glyphName, transform, self.glyphSet.rawLocation
+                        )
+                    except AttributeError:
+                        t = transform.toTransform()
+                        compGlyphSet = (
+                            self.glyphSet
+                            if comp.glyphName != self.name
+                            else glyphSet.glyphSet
+                        )
```
Comment (anthrotype) [inline Lib/fontTools/ttLib/ttGlyphSet.py:337]: is this the part where you guard against the infinite recursion?

https://github.com/harfbuzz/boring-expansion-spec/blob/main/VARC.md#processing

> if a component refers to the glyphID of the current glyph, instead of recursing (which would result in infinite recursion), the glyph outline for the component is loaded directly instead, as if the glyphID had no VARC entry.

Isn't it possible that a cycle occurs not immediately from current glyph to a direct component, but indirectly through some additional steps in the chain of component references?
E.g. a varcomposite glyph reference some varcomponent, whose base glyph is in turn a varcomposite which has a varcomponent referencing the original varcomposite, etc...

Comment (behdad) [reply Lib/fontTools/ttLib/ttGlyphSet.py:337]: This is NOT to prevent a cycle. Cycle detection needs to be implemented separately. I don't think we do it for the regular components either.

This is specifically to allow mixing outlines with components, but allowing a VARC glyphid to refer to its own glyf glyphid.

**Comment by behdad** [comment]
Okay. I'll take care of flags now.

**Comment by behdad** [comment]
> Okay. I'll take care of flags now.

Done. I still read an optional `flags` tag...

**Comment by behdad** [comment]
Currently I'm outputing:
```
          <resetUnspecifiedAxes value="True"/>
```
for the lone necessary flag only if it's set. Should I remove the `value` attribute from it?

**Comment by anthrotype** [comment]
> Done. I still read an optional flags tag...

so you read it but don't write it out any more? what's the use of that?

> Currently I'm outputing: <resetUnspecifiedAxes value="True"/> for the lone necessary flag only if it's set. Should I remove the value attribute from it?

As in attribute-less element i.e. `<resetUnspecifiedAxes/>`? How about you write `resetUnspecifiedAxes="True"` instead as an actual xml _attribute_ of the `VarComponent` element, or actually closer to where it's being used, an attribute of `axisValues` element?

File: Lib/fontTools/ttLib/ttGlyphSet.py (hunk 6 above)
Comment (anthrotype) [reply Lib/fontTools/ttLib/ttGlyphSet.py:337]: gotcha

**Comment by anthrotype** [comment]
note that for a similar boolean flag elsewhere we write `"1"` instead of `"True"` in the xml dump, you may want to do the same for `resetUnspecifiedAxes`:

https://github.com/fonttools/fonttools/blob/679a5c0c252fd5b983de0266b0a250fc7cbb49fc/Lib/fontTools/ttLib/tables/_g_l_y_f.py#L788-L789

**Comment by behdad** [comment]
> > Done. I still read an optional flags tag...
> 
> so you read it but don't write it out any more? what's the use of that?

Nothing honestly. I'll remove it.

> > Currently I'm outputing:  for the lone necessary flag only if it's set. Should I remove the value attribute from it?
> 
> As in attribute-less element i.e. `<resetUnspecifiedAxes/>`? How about you write `resetUnspecifiedAxes="True"` instead as an actual xml _attribute_ of the `VarComponent` element, or actually closer to where it's being used, an attribute of `axisValues` element?

Okay I'll add it to `axisValues`, and make it output 1. Default to 0? Or always output it?

**Comment by anthrotype** [comment]
I'd say default to 0 and only write if 1

**Comment by behdad** [comment]
XML output for `resetUnspecifiedAxes` adjusted as per feedback. Reading of `flags` removed.

**Comment by behdad** [comment]
HarfBuzz counterpart WIP: https://github.com/harfbuzz/harfbuzz/pull/4578

**Comment by behdad** [comment]
> HarfBuzz counterpart WIP: [harfbuzz/harfbuzz#4578](https://github.com/harfbuzz/harfbuzz/pull/4578)

Now can render fonts generated with this fonttools branch.

**Comment by behdad** [comment]
This is ready. Has tests too.

File: Lib/fontTools/ttLib/tables/TupleVariation.py [hunk 7]
```diff
@@ -22,6 +22,8 @@
 
 DELTAS_ARE_ZERO = 0x80
 DELTAS_ARE_WORDS = 0x40
+DELTAS_ARE_LONGS = 0xC0
```
[the remark this answers is not in the document]
Comment (behdad) [reply Lib/fontTools/ttLib/tables/TupleVariation.py:25]: You caught me!

I'll see if I can make one.

[the remark this answers is not in the document]
Comment (anthrotype) [reply Lib/fontTools/ttLib/tables/TupleVariation.py:25]: > You caught me!

codecov did! :)

**Comment by justvanrossum** [comment]
Suggest to apply this patch to make a var composite referencing its own GID work correctly with the pen protocol:

```diff
diff --git a/Lib/fontTools/ttLib/ttGlyphSet.py b/Lib/fontTools/ttLib/ttGlyphSet.py
index 8b276aa37..5d373cfb9 100644
--- a/Lib/fontTools/ttLib/ttGlyphSet.py
+++ b/Lib/fontTools/ttLib/ttGlyphSet.py
@@ -344,11 +344,17 @@ class _TTGlyphVARC(_TTGlyph):
             reset = comp.flags & VarComponentFlags.RESET_UNSPECIFIED_AXES
             with self.glyphSet.glyphSet.pushLocation(location, reset):
                 with self.glyphSet.pushLocation(location, reset):
-                    try:
-                        pen.addVarComponent(
-                            comp.glyphName, transform, self.glyphSet.rawLocation
-                        )
-                    except AttributeError:
+                    shouldDecompose = self.name == comp.glyphName
+
+                    if not shouldDecompose:
+                        try:
+                            pen.addVarComponent(
+                                comp.glyphName, transform, self.glyphSet.rawLocation
+                            )
+                        except AttributeError:
+                            shouldDecompose = True
+
+                    if shouldDecompose:
                         t = transform.toTransform()
                         compGlyphSet = (
                             self.glyphSet
```

Without this, it will emit a self-referencing component which (as far as I can see) can't be resolved in another way.

@behdad, I can't find a reference to the prescribed behavior in the [proposal doc](https://github.com/harfbuzz/boring-expansion-spec/blob/main/VARC.md).

Others: what was new to me is that a VARC Composite may reference its own GID, which will then bypass VARC and go straight to glyf or CFF2. This is a handy way to mix outlines and (variable) components.

**Comment by behdad** [comment]
@justvanrossum Thanks. Can you push that change to the branch?

> @behdad, I can't find a reference to the prescribed behavior in the [proposal doc](https://github.com/harfbuzz/boring-expansion-spec/blob/main/VARC.md).

Here:

  The component then is recursively loaded from the VARC table, if it is present in this table; otherwise, falls back to glyf/gvar, CFF2, or any other mechanism to get raw outlines. An exception to the recursive rule is that if a component refers to the glyphID of the current glyph, instead of recursing (which would result in infinite recursion), the glyph outline for the component is loaded directly instead, as if the glyphID had no VARC entry.

**Comment by behdad** [comment]
I suggest landing this, as it's on the standardization pipeline now and settled.

**Comment by justvanrossum** [comment]
A release would be the cherry on the cake.

**Comment by anthrotype** [comment]
I can do that

**Comment by anthrotype** [comment]
pushed tag https://github.com/fonttools/fonttools/releases/tag/4.52.0, wheels are building..

big release, perhaps worth a tweet?

**Comment by behdad** [comment]
> pushed tag https://github.com/fonttools/fonttools/releases/tag/4.52.0, wheels are building..
> 
> big release, perhaps worth a tweet?

Yes please.

In HB we're thinking about calling it a 9.0.0

**Comment by anthrotype** [comment]
in fonttools we bump major very infrequently, only when we break backward compatibility (e.g. when we swichted from py2.py3 to py3 only)
